Bluetooth: Controller: Ext Adv Auxiliary PDUs with ticks_slot_window

cvinayak opened the pull request:
Add implementation for Extended Advertising Auxiliary PDUs
to use ticks slot window feature.

This will allow the periodic scheduling of AUX_ADV_IND PDUs
to drift upto its next interval when overlapping with other
states/roles that cannot be moved around, to avoid skipping
them.

Having an active Extended Advertising simultaneously with
an ISO Synchronized Receiver or Connected ISO connection
will now have less ISO SDU loss.

The feature is a one-line assignment to use, but a conditional
compilation mess! Lets address that when it comes to #65125

cvinayak committed:
Bluetooth: Controller: Ext Adv Auxiliary PDUs with ticks_slot_window

Add implementation for Extended Advertising Auxiliary PDUs
to use ticks slot window feature.

This will allow the periodic scheduling of AUX_ADV_IND PDUs
to drift upto 10 ms advertising delay minus the ticks_slot
time reservation of the AUX_ADV_IND PDU when overlapping
with other states/roles that cannot be moved around, to
avoid skipping them.

Having an active Extended Advertising simultaneously with
an ISO Synchronized Receiver or Connected ISO connection
will now have less ISO SDU loss when using 10 ms ISO
intervals.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Introduce ticker reschedule with drift

Introduce ticker reschedule with drift so that role like
AUX_ADV_IND can start after overlapping states and roles
using time reservations.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix to reschedule before overlap when yielding

Fix to reschedule before overlap and be collision resolved
in the next periodic interval for tickers using slot window
yield.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: hci_ipc: Fix Newton's Cradle, reduce RAM usage

The required ISO Tx buffers have to match the Read Buffer
Size values, otherwise the difference in the value cause a
similar amount of buffers to be stalled in the IPC driver.

Use reduced HEAP size to make room for increase in RAM
usage.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/hci_ipc/overlay-nrf5340_cpunet_iso_nrf21540_ek-bt_ll_sw_split.conf:
Hmm, this looks a bit messy, and the order of the .conf files will affect the final result, right? 

Not sure this is the best way to do achieve this.

cvinayak replied:
This is a single `EXTRA_CONF_FILE=`, I believe  is merged after the `CONF_FILE`

```
Parsing /home/jack/workspace/zephyrproject/zephyr/Kconfig
Loaded configuration '/home/jack/workspace/zephyrproject/zephyr/boards/nordic/nrf5340_audio_dk/nrf5340_audio_dk_nrf5340_cpunet_defconfig'
Merged configuration '/home/jack/workspace/zephyrproject/zephyr/samples/bluetooth/hci_ipc/nrf5340_cpunet_iso-bt_ll_sw_split.conf'
Merged configuration '/home/jack/workspace/zephyrproject/zephyr/samples/bluetooth/hci_ipc/overlay-nrf5340_cpunet_iso_nrf21540_ek-bt_ll_sw_split.conf'
Configuration saved to '/home/jack/workspace/zephyrproject/zephyr/build/hci_ipc/zephyr/.config'

```

Thalley replied:
> This is a single `EXTRA_CONF_FILE=`, I believe is merged after the `CONF_FILE`
> 
> ```
> Parsing /home/jack/workspace/zephyrproject/zephyr/Kconfig
> Loaded configuration '/home/jack/workspace/zephyrproject/zephyr/boards/nordic/nrf5340_audio_dk/nrf5340_audio_dk_nrf5340_cpunet_defconfig'
> Merged configuration '/home/jack/workspace/zephyrproject/zephyr/samples/bluetooth/hci_ipc/nrf5340_cpunet_iso-bt_ll_sw_split.conf'
> Merged configuration '/home/jack/workspace/zephyrproject/zephyr/samples/bluetooth/hci_ipc/overlay-nrf5340_cpunet_iso_nrf21540_ek-bt_ll_sw_split.conf'
> Configuration saved to '/home/jack/workspace/zephyrproject/zephyr/build/hci_ipc/zephyr/.config'
> ```

Right, but this isn't very clear to users. The order of the configuration files is not really documented to be in any specific order, and they are not applied automatically correctly, so this can easily cause issues for users I believe.

It can be OK if we document it in the sample

@kartben may also have some input here

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -71,11 +71,18 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t ticks_drift,
 static void *adv_aux_free;
 
-#if defined(CONFIG_BT_CTLR_ADV_PERIODIC) && defined(CONFIG_BT_TICKER_EXT_EXPIRE_INFO)
+#if defined(CONFIG_BT_CTLR_ADV_AUX_SLOT_WINDOW) || \
+	(defined(CONFIG_BT_CTLR_ADV_PERIODIC) && \
+	 defined(CONFIG_BT_TICKER_EXT_EXPIRE_INFO))
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC) && \
+	defined(CONFIG_BT_TICKER_EXT_EXPIRE_INFO)
```
Man, this some tasty `defined` spaghetti :spaghetti: 

I assume that it required given the existing dependencies on `#if defined` in the controller, where nearly everything is guarded one way or another. 

At some point it would be interesting to start removing these defined for all prototypes (both static and the ones in header files), so we can start using `IS_ENABLED` more.

cvinayak replied:
Unfortunately, ISO feature scheduling requirements have been challenging, and it has been difficult to avoid regressions in behavior change without conditionally compiling out new implementations.

Thalley commented on samples/bluetooth/hci_ipc/nrf5340_cpunet_bis-bt_ll_sw_split.conf:
```diff
@@ -72,6 +72,7 @@ CONFIG_BT_CTLR_ADV_DATA_BUF_MAX=6
 CONFIG_BT_CTLR_SCAN_AUX_SET=1
 
 CONFIG_BT_CTLR_ADV_RESERVE_MAX=n
+CONFIG_BT_CTLR_ADV_AUX_SLOT_WINDOW=y
```
Since this is being enable in so many places, should it be enabled by default?

cvinayak replied:
Realized that it has behavioral changes to non-ISO usecases when enabled by default. As ISO features in the Controller are experimental, it is best to use this feature only with ISO builds.

Thalley replied:
We can default it for ISO then?

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -437,6 +437,14 @@ config BT_CTLR_ADV_RESERVE_MAX
 	  corresponding to the Advertising Data present at the time of the
 	  start/enable of Advertising is used.
 
+config BT_CTLR_ADV_AUX_SLOT_WINDOW
+	bool "Reschedule Extended Advertising Auxiliary PDUs"
```
Name and description doesn't seem to match

Thalley commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -946,21 +946,23 @@ static uint8_t ticker_resolve_collision(struct ticker_node *nodes,
+					 TICKER_HAS_SLOT_WINDOW(ticker)) ||
+					((acc_ticks_to_expire +
+					  ticker_next->ext_data->ticks_slot_window -
+					  ticker_next->ticks_slot) <
+					 ticker->ticks_slot);
 
 			/* Can the current ticker with ticks_slot_window be
 			 * scheduled after the colliding ticker?
 			 */
 			uint8_t curr_has_ticks_slot_window =
-					(TICKER_HAS_SLOT_WINDOW(ticker) &&
-					 ((acc_ticks_to_expire +
-					   ticker_next->ticks_slot) <
-					  (ticker->ext_data->ticks_slot_window -
-					   ticker->ticks_slot)));
+					TICKER_HAS_SLOT_WINDOW(ticker) &&
+					((acc_ticks_to_expire +
+					  ticker_next->ticks_slot) <=
+					 (ticker->ext_data->ticks_slot_window -
+					  ticker->ticks_slot));
```
Several of these calculations are getting quite complex. Should functions be implemented to help calculate `next_not_ticks_slot_window` and `curr_has_ticks_slot_window`?

The worst part of these is that the calculations are undocumented: It's not clear to me why they are calculated as such, and are hard to maintain.

cvinayak replied:
Updated the rules for collision resolution which are documented in the function description.

The logical expression at line 1012-1019 reflect the documented rules, the variables are accordingly named to reflect the rule sentences.

Thalley commented on subsys/bluetooth/controller/ticker/ticker.h:
```diff
@@ -229,15 +229,24 @@ uint8_t ticker_priority_set(uint8_t instance_index, uint8_t user_id,
+					* is_jitter_in_window, otherwise is only
+					* the value of drift due to reschedule
+					*/
+	uint8_t reschedule_state:3;    /* State of re-scheduling of the
+					* node. See defines
+					* TICKER_RESCHEDULE_STATE_XXX
+					*/
```
If a new value is added to `TICKER_RESCHEDULE_STATE_` then this will silently start failing, right? 

How can we prevent a new `TICKER_RESCHEDULE_STATE_` value to fail silently? Add comment to the `TICKER_RESCHEDULE_STATE_` defines? BUILD_ASSERT?

cvinayak replied:
BUILD_ASSERT will be best... will investigate on how to add the check.

kruithofa replied:
I think we already get a compiler warning if the number of states becomes larger than what we can store in `reschedule_state`

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -437,6 +437,14 @@ config BT_CTLR_ADV_RESERVE_MAX
 	  start/enable of Advertising is used.
 
+config BT_CTLR_ADV_AUX_SLOT_WINDOW
+	bool "Reschedule Extended Advertising Auxiliary PDUs"
```
Is this ok?

```suggestion
	bool "Reschedule Extended Advertising Auxiliary PDUs within a slot window range"
```

Thalley replied:
The name `BT_CTLR_ADV_AUX_SLOT_WINDOW` does not match "reschedule" IMO. From someone not as familiar with the controller as you, the description and help text looks more like an accidental copy-and-paste from a different Kconfig, than a description of this Kconfig. 

Perhaps `resched` should be part of the name?

cvinayak committed:
tests: bsim: Bluetooth: Fix simulation length be greater than wait time

Fix some test simulation length be greater than wait time.
These tests completed within 12 seconds, 15 second wait
time is a good value with included margin and a 20 second
simulation time.

These tests failed on this PR CI.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley reviewed (commented):
Would like a reply on https://github.com/zephyrproject-rtos/zephyr/pull/80733#discussion_r1836124059

Thalley commented on samples/bluetooth/hci_ipc/nrf5340_cpunet_bt_mesh-bt_ll_sw_split.conf:
```diff
@@ -1,7 +1,7 @@
 CONFIG_MBOX=y
 
-CONFIG_HEAP_MEM_POOL_SIZE=8192
+CONFIG_HEAP_MEM_POOL_SIZE=6144
```
Do we need this reduction for mesh?

Thalley commented on samples/bluetooth/hci_ipc/nrf5340_cpunet_cis-bt_ll_sw_split.conf:
```diff
@@ -5,7 +5,7 @@ CONFIG_ISR_STACK_SIZE=1024
 CONFIG_MAIN_STACK_SIZE=512
 CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=512
-CONFIG_HEAP_MEM_POOL_SIZE=8192
+CONFIG_HEAP_MEM_POOL_SIZE=6144
```
Do we need this reduction for all ISO .confs? 

Isn't it only the combined one? 
Or is this just done to be consistent for all of them?

Thalley commented on samples/bluetooth/hci_ipc/nrf5340_cpunet_df-bt_ll_sw_split.conf:
```diff
@@ -5,7 +5,7 @@ CONFIG_ISR_STACK_SIZE=1024
 CONFIG_MAIN_STACK_SIZE=512
 CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=512
-CONFIG_HEAP_MEM_POOL_SIZE=8192
+CONFIG_HEAP_MEM_POOL_SIZE=6144
```
Ditto, do we need this reduction for DF?

Thalley commented on samples/bluetooth/hci_ipc/nrf5340_cpunet_bis-bt_ll_sw_split.conf:
```diff
@@ -40,7 +40,11 @@ CONFIG_BT_ISO_PERIPHERAL=n
 # ISO Streams
 CONFIG_BT_ISO_MAX_CHAN=4
-CONFIG_BT_ISO_TX_BUF_COUNT=1
+# In theory, CONFIG_BT_ISO_TX_BUF_COUNT=1, should be sufficient but this count
+# is used in the context of IPC which falls into a "Newton's Cradle" effect
+# where probably (CONFIG_BT_CTLR_ISO_TX_BUFFERS - CONFIG_BT_ISO_TX_BUF_COUNT)
+# buffers get throttled. Hence, always have the value equal or greater.
+CONFIG_BT_ISO_TX_BUF_COUNT=12
```
Could this be further explained in a GH issue?

This isn't something I understand as it is written here

Thalley commented on samples/bluetooth/hci_ipc/overlay-nrf5340_cpunet_iso_nrf21540_ek-bt_ll_sw_split.conf:
```diff
@@ -0,0 +1,6 @@
+# Reduce from 310 bytes, in nrf5340_cpunet_iso-bt_ll_sw_split.conf
+# to be able to fit in 64KB RAM, in case needed in the future.
+
+# Example:
+# CONFIG_BT_ISO_TX_MTU=247
+# CONFIG_BT_ISO_RX_MTU=251
```
As I mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/80733#discussion_r1830602527: This is messy. We shouldn't have overlays/extra conf files that need to be added in a specific order to work, especially when it's not documented.

Thalley replied:
> OVERLAY_CONFIG is deprecated, but it did convey that something was overlay-ed!

Right :D I should stop calling them overlays

> There is repo wide use of extra_configs in sample.yaml and testcase.yaml files to use overrides of Kconfig values from prj.conf and use of EXTRA_CONF_FILE is better organization wise compared to use of list of Kconfigs under extra_configs.

If we have several places where we use multiple extra configs that needs to be specific in a specific order, that's not a good reason to do it here either. IMO that just show that we have this other places too. 

A file called `testcase.yaml` does/should not be used as documentation. If these values are required for this specific board to work, then that needs to be documented and/or applied automatically.

cvinayak replied:
It is documented  in the snippets section that the processing order of the snippets, `.conf` and `.overlay` files are "in the order they are listed".

https://github.com/zephyrproject-rtos/zephyr/blob/28796076d84056d085e5a4ebf213dafbfe56202a/doc/build/snippets/writing.rst?plain=1#L98-L126

Thalley commented on tests/bsim/bluetooth/audio/overlay-bt_ll_sw_split.conf:
```diff
@@ -1,45 +1,119 @@
-# Controller Settings
-CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=255
+# Host and Controller common dependencies
+CONFIG_BT_BROADCASTER=y
+CONFIG_BT_OBSERVER=y
+CONFIG_BT_EXT_ADV=y
+CONFIG_BT_PER_ADV=y
+CONFIG_BT_PER_ADV_SYNC=y
+CONFIG_BT_PER_ADV_SYNC_MAX=2
+CONFIG_BT_CENTRAL=y
+CONFIG_BT_PERIPHERAL=y
+CONFIG_BT_MAX_CONN=3
+
+# Broadcast and Connected ISO
+CONFIG_BT_ISO_BROADCASTER=y
+CONFIG_BT_ISO_SYNC_RECEIVER=y
+CONFIG_BT_ISO_CENTRAL=y
+CONFIG_BT_ISO_PERIPHERAL=y
```
Since this is an overlay, what's the purpose or overlaying these values when they are already provided by the "main" conf file? 

Shouldn't the overlay only contain values that are different?

cvinayak replied:
I have removed the commit that makes the nrf52 same as nrf53... will address that in a separate PR, may be having different Kconfig for nRF52 and nRF53 could be better in long run to discover bugs.

cvinayak commented:
#AutoPTS run zephyr nrf53 BAP/UCL/STR/BV-523-C

Thalley commented:
#AutoPTS run zephyr nrf53 BAP/UCL/STR/BV-523-C

codecoup-tester commented:
AutoPTS Bot results:
No failed test found.
<details><summary>Successful tests (1)</summary>BAP     BAP/UCL/STR/BV-523-C            PASS (2)</details>
